src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala

```diff
@@ -0,0 +1,89 @@
+  */
+class ManagedGroupService(resourceService: ResourceService, val resourceTypes: Map[ResourceTypeName, ResourceType]) extends LazyLogging {
+
+  def managedGroupType: ResourceType = resourceTypes.getOrElse(ManagedGroupService.ManagedGroupTypeName, throw new WorkbenchException(s"resource type ${ManagedGroupService.ManagedGroupTypeName.value} not found"))
+  def memberRole = managedGroupType.roles.find(_.roleName == ManagedGroupService.MemberRoleName).getOrElse(throw new WorkbenchException(s"${ManagedGroupService.MemberRoleName} role does not exist in $managedGroupType"))
+
+  // Important Note, see: org/broadinstitute/dsde/workbench/sam/model/SamModel.scala:90
+  def createManagedGroup(groupId: ResourceId, userInfo: UserInfo): Future[Resource] = {
+    for {
+      managedGroup <- resourceService.createResource(managedGroupType, groupId, userInfo)
+      _ <- createPolicyForMembers(managedGroup)
+      policies <- resourceService.accessPolicyDAO.listAccessPolicies(managedGroup)
+      _ <- createAggregateGroup(managedGroup, policies)
+    } yield managedGroup
+  }
+
+  private def createPolicyForMembers(managedGroup: Resource): Future[AccessPolicy] = {
+    val accessPolicyName = AccessPolicyName(memberRole.roleName.value)
+    val resourceAndPolicyName = ResourceAndPolicyName(managedGroup, accessPolicyName)
+    resourceService.createPolicy(resourceAndPolicyName, members = Set.empty, Set(memberRole.roleName), actions = Set.empty)
+  }
+
+  private def createAggregateGroup(resource: Resource, componentPolicies: Set[AccessPolicy]): Future[BasicWorkbenchGroup] = {
+    val email = generateManagedGroupEmail(resource.resourceId)
+    val workbenchGroupName = WorkbenchGroupName(resource.resourceId.value)
+    val groupMembers: Set[WorkbenchSubject] = componentPolicies.map(_.id)
+    resourceService.directoryDAO.createGroup(BasicWorkbenchGroup(workbenchGroupName, groupMembers, email))
+  }
+
+  private def generateManagedGroupEmail(resourceId: ResourceId): WorkbenchEmail = {
+    val localPart = resourceId.value
+    validateEmail(localPart)
+    WorkbenchEmail(constructEmail(localPart))
+  }
+
+  private def constructEmail(localPart: String) = {
+    s"${localPart}@${resourceService.emailDomain}"
+  }
+
+  private def validateEmail(localPart: String) = {
+    val errors = validateLocalPartPattern(localPart) ++ validateEmailLength(constructEmail(localPart))
+    if (errors.nonEmpty)
+      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot create valid email address from group name" , errors.toSeq))
+  }
+
+  private def validateLocalPartPattern(str: String): Option[ErrorReport] = {
+    ManagedGroupService.LocalPartRe.findFirstMatchIn(str) match {
+      case Some(_) => None
+      case None => Option(ErrorReport(s"You have specified a group name that contains characters that are not permitted in an email address. Group name may only contain alphanumeric characters, underscores, and dashes"))
+    }
```

Comment:
regexes can be used in pattern matches so this also works

```
    str match {
      case ManagedGroupService.LocalPartRe() => None
      case _ => Option(ErrorReport(s"You have specified a group name that contains characters that are not permitted in an email address. Group name may only contain alphanumeric characters, underscores, and dashes")) 
    }
```

Reply:
fixed